src/compas_fab/datastructures/plan.py

```diff
@@ -0,0 +1,371 @@
+    @staticmethod
+    def compose_message(invalid_ids, pa_id):
+        if pa_id:
+            return 'Planned action {} has invalid dependency ids {}'.format(pa_id, invalid_ids)
+        return 'Found invalid dependency ids {}'.format(invalid_ids)
+
+
+class Plan(Datastructure):
+    """Data structure for holding the information of a partially ordered plan
+    (a directed acyclic graph).  The content of any event of the plan is contained
+    in an :class:`compas_fab.datastructures.Action`. An event is scheduled and
+    added to the plan through a :class:`compas_fab.datastructures.PlannedAction`.
+    The dependency ids of a planned action can be thought of as pointers to the
+    parents of that planned action.
+
+    Parameters
+    ----------
+    planned_actions : list of :class:`compas_fab.datastructures.PlannedAction`
+        The planned actions will be stored as an ordered dictionary, so their
+        ids should be distinct.  Defaults to an empty list.
+    id_generator : Generator[Hashable, None, None]
+        Object which generates keys (via ``next()``) for
+        :class:`compas_fab.datastructures.Action`s added using this object's
+        methods.  Defaults to :class:`compas_fab.datastructures.IntegerIdGenerator`.
+    """
+    def __init__(self, planned_actions=None, id_generator=None):
+        super(Plan, self).__init__()
+        planned_actions = planned_actions or []
+        self.planned_actions = planned_actions
+        if id_generator is None:
+            try:
+                start_value = max(self.planned_actions.keys()) + 1 if self.planned_actions else 1
+            except Exception:
+                raise Exception('Given ids not compatible with default id_generator.')
+            id_generator = IntegerIdGenerator(start_value)
+        self._id_generator = id_generator
+
+    @property
+    def planned_actions(self):
+        return self._planned_actions
+
+    @planned_actions.setter
+    def planned_actions(self, planned_actions):
+        self._planned_actions = OrderedDict({pa.id: pa for pa in planned_actions})
+        self.check_all_dependency_ids()
+
+    def plan_action(self, action, dependency_ids):
+        """Adds the action to the plan with the given dependencies,
+        and generates an id for the newly planned action.
+
+        Parameters
+        ----------
+        action : :class:`comaps_fab.datastructures.Action`
+            The action to be added to the plan.
+        dependency_ids : set or list
+            The keys of the already planned actions that the new action
+            is dependent on.
+
+        Returns
+        -------
+            The id of the newly planned action.
+        """
+        self.check_dependency_ids(dependency_ids)
+        action_id = self._get_next_action_id()
+        planned_action = PlannedAction(action_id, action, dependency_ids)
+        self.planned_actions[action_id] = planned_action
+        return action_id
+
+    def append_action(self, action):
```

Comment:
I wanted to make a bigger distinction between `dependency_ids=None` and `dependency_ids=[]` because they result in very different behavior and I can foresee that being the source of annoying bugs for the user.